### PR TITLE
Feature broker 23: Shared subscriptions

### LINKED ISSUE #23: Implement shared subscriptions
Implement shared subscriptions

### PR DESCRIPTION
**Pull request by crazyrokr**

### CODE REVIEW

File: src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+@RequiredArgsConstructor
+public class SharedSubscriber {
+
+    private final @Getter @NotNull String group;
+    private final Array<Subscriber> subscribers;
+    private final StampedLock lock;
+    private int current;
+
+    public SharedSubscriber(@NotNull Subscriber subscriber) {
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java]: why do we have a subscriber as constructor an argument? I think it should be the only a group name.

File: src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.Objects;
+import java.util.concurrent.locks.StampedLock;
+
+@RequiredArgsConstructor
+public class SharedSubscriber {
+
+    private final @Getter @NotNull String group;
+    private final Array<Subscriber> subscribers;
+    private final StampedLock lock;
+    private int current;
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java]: it should be atomic integer

File: src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+import java.util.Objects;
+import java.util.concurrent.locks.StampedLock;
+
+@RequiredArgsConstructor
+public class SharedSubscriber {
+
+    private final @Getter @NotNull String group;
+    private final Array<Subscriber> subscribers;
+    private final StampedLock lock;
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java]: we don't need to have an additional lock here

File: src/main/java/com/ss/mqtt/broker/model/Subscriber.java [hunk 4]
```diff
@@ -33,4 +26,15 @@ public Subscriber(@NotNull MqttClient mqttClient, @NotNull SubscribeTopicFilter
         return subscribeTopicFilter.getTopicFilter();
     }
 
+    public boolean isShared() {
+        return TopicFilter.isShared(subscribeTopicFilter.getTopicFilter());
+    }
+
+    public @NotNull String getGroup() {
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/Subscriber.java]: why do you not use different types of subscriber class to understand shared/or not?

File: src/main/java/com/ss/mqtt/broker/model/topic/AbstractTopic.java [hunk 5]
```diff
@@ -19,29 +19,28 @@
     static final String MULTI_LEVEL_WILDCARD = "#";
     static final String SINGLE_LEVEL_WILDCARD = "+";
 
-    static void checkTopic(@NotNull String topic) {
-        if (topic.length() == 0) {
-            throw new IllegalArgumentException("Topic has zero length.");
-        } else if (topic.contains("//") || topic.startsWith("/") || topic.endsWith("/")) {
-            throw new IllegalArgumentException("Topic has zero length level: " + topic);
-        }
+    static boolean checkTopic(@NotNull String topic) {
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/topic/AbstractTopic.java]: I think it should be in some utils class

File: src/main/java/com/ss/mqtt/broker/model/topic/AbstractTopic.java [hunk 6]
```diff
@@ -19,29 +19,28 @@
...
+        return topic.length() != 0 && !topic.contains("//") && !topic.startsWith("/") && !topic.endsWith("/");
     }
 
     private final @NotNull String[] segments;
     private final @NotNull String rawTopic;
+    private final @Getter boolean invalid;
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/topic/AbstractTopic.java]: why do we need to have an additional flag in this class to understand that this topic is not valid?

File: src/main/java/com/ss/mqtt/broker/model/topic/TopicSubscribers.java [hunk 7]
```diff
@@ -58,19 +85,37 @@ private static boolean removeDuplicateWithLowerQoS(
...
+            } finally {
+                subscribers.readUnlock(stamp);
+            }
+        }
     }
 
     private volatile @Getter @Nullable ConcurrentObjectDictionary<String, TopicSubscribers> topicSubscribers;
-    private volatile @Getter @Nullable ConcurrentArray<Subscriber> subscribers;
+    private volatile @Getter @Nullable ConcurrentObjectDictionary<String, SharedSubscriber> sharedSubscribers;
+    private volatile @Getter @Nullable ConcurrentArray<Subscriber> singleSubscribers;
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/topic/TopicSubscribers.java]: I think we can reuse singleSubscribers for shared subscribers as well

**Review by JavaSaBr — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/mqtt/broker/handler/packet/in/DisconnetInPacketHandler.java [hunk 8]
```diff
@@ -20,6 +20,6 @@ protected void handleImpl(@NotNull UnsafeMqttClient client, @NotNull DisconnectI
             log.error("Disconnect client {} by error reason {}", client, reasonCode);
         }
 
-        client.release().subscribe();
+        client.getConnection().close();
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/handler/packet/in/DisconnetInPacketHandler.java:23]: why did you change release to close?

File: src/main/java/com/ss/mqtt/broker/handler/packet/in/SubscribeInPacketHandler.java [hunk 9]
```diff
@@ -15,5 +19,14 @@
...
         var ackReasonCodes = subscriptionService.subscribe(client, packet.getTopicFilters());
         client.send(client.getPacketOutFactory().newSubscribeAck(packet.getPacketId(), ackReasonCodes));
+        ackReasonCodes.forEach(client, SubscribeInPacketHandler::checkException);
+    }
+
+    private static void checkException(@NotNull UnsafeMqttClient client, @NotNull SubscribeAckReasonCode reason) {
+        if (reason == SHARED_SUBSCRIPTIONS_NOT_SUPPORTED || reason == WILDCARD_SUBSCRIPTIONS_NOT_SUPPORTED) {
+            int reasonIndex = Byte.toUnsignedInt(reason.getValue());
+            var disconnect = client.getPacketOutFactory().newDisconnect(client, DisconnectReasonCode.of(reasonIndex));
+            client.sendWithFeedback(disconnect).thenAccept(result -> client.getConnection().close());
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/handler/packet/in/SubscribeInPacketHandler.java:34]: why do you use close instead of release?

File: src/main/java/com/ss/mqtt/broker/handler/packet/in/SubscribeInPacketHandler.java [hunk 10]
```diff
@@ -15,5 +19,14 @@
     protected void handleImpl(@NotNull UnsafeMqttClient client, @NotNull SubscribeInPacket packet) {
         var ackReasonCodes = subscriptionService.subscribe(client, packet.getTopicFilters());
         client.send(client.getPacketOutFactory().newSubscribeAck(packet.getPacketId(), ackReasonCodes));
+        ackReasonCodes.forEach(client, SubscribeInPacketHandler::checkException);
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/handler/packet/in/SubscribeInPacketHandler.java]: if you have for example two SHARED_SUBSCRIPTIONS_NOT_SUPPORTED reason codes, you will send a disconnect packet twice

File: src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+import com.ss.rlib.common.util.array.ConcurrentArray;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class SharedSubscriber implements Subscriber {
+
+    private final SharedTopicFilter topicFilter;
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java]: missed annotations

File: src/main/java/com/ss/mqtt/broker/model/topic/TopicSubscribers.java [hunk 12]
```diff
@@ -18,27 +19,137 @@
...
+
+        var singleSubscriber = new SingleSubscriber(client, subscribe);
+        sharedSubscriber.addSubscriber(singleSubscriber);
+    }
+
+    private static boolean sharedSubscriberWithGroup(@NotNull String group, @NotNull Subscriber subscriber) {
+        return isSharedSubscriber(subscriber) && group.equals(((SharedSubscriber) subscriber).getGroup());
+    }
+
+    private static MqttClient singleSubscriberToMqttClient(@NotNull Subscriber subscriber) {
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/topic/TopicSubscribers.java]: missed annotation

**Review by JavaSaBr — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/mqtt/broker/handler/packet/in/DisconnetInPacketHandler.java (hunk 8 above)
Comment (crazyrokr) [reply src/main/java/com/ss/mqtt/broker/handler/packet/in/DisconnetInPacketHandler.java:23]: Because of deduplication

File: src/main/java/com/ss/mqtt/broker/handler/packet/in/SubscribeInPacketHandler.java (hunk 9 above)
Comment (crazyrokr) [reply src/main/java/com/ss/mqtt/broker/handler/packet/in/SubscribeInPacketHandler.java:34]: Because of deduplication

File: src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java (hunk 11 above)
Comment (crazyrokr) [reply src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java]: Is it really required ?
Method com.ss.mqtt.broker.model.SubscribeTopicFilter#getTopicFilter returns NotNull value.

File: src/main/java/com/ss/mqtt/broker/handler/packet/in/DisconnetInPacketHandler.java (hunk 8 above)
Comment (JavaSaBr) [reply src/main/java/com/ss/mqtt/broker/handler/packet/in/DisconnetInPacketHandler.java:23]: but when you close this connection, you will not release its client

File: src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class SharedSubscriber implements Subscriber {
+
+    private final @NotNull SharedTopicFilter topicFilter;
+    private final ConcurrentArray<SingleSubscriber> subscribers;
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java]: missed annotations

File: src/main/java/com/ss/mqtt/broker/util/TopicUtils.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+import org.jetbrains.annotations.NotNull;
+
+public class TopicUtils {
+
+    private static final TopicFilter INVALID_TOPIC_FILTER = new TopicFilter();
+    private static final TopicName INVALID_TOPIC_NAME = new TopicName();
+    public static final TopicName EMPTY_TOPIC_NAME = new TopicName();
+
+    private static final String SHARE_KEYWORD = "$share";
+    public static final String DELIMITER = "/";
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/util/TopicUtils.java]: I think these fields should be private

**Review by JavaSaBr — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/mqtt/broker/model/SingleSubscriber.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
+package com.ss.mqtt.broker.model;
+
+import com.ss.mqtt.broker.network.client.MqttClient;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+import org.jetbrains.annotations.NotNull;
+
+@ToString
+@EqualsAndHashCode(of = "mqttClient", callSuper = false)
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/SingleSubscriber.java]: callSuper is false by default

File: src/main/java/com/ss/mqtt/broker/handler/packet/in/SubscribeInPacketHandler.java [hunk 16]
```diff
@@ -15,5 +19,17 @@
     protected void handleImpl(@NotNull UnsafeMqttClient client, @NotNull SubscribeInPacket packet) {
         var ackReasonCodes = subscriptionService.subscribe(client, packet.getTopicFilters());
         client.send(client.getPacketOutFactory().newSubscribeAck(packet.getPacketId(), ackReasonCodes));
+        ackReasonCodes.findAny(client, SubscribeInPacketHandler::hasInvalidReason);
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/handler/packet/in/SubscribeInPacketHandler.java]: you should take the 'ackCode' by condition and if it's not null then close this client inside this method, it's not correct to do it inside the search condition.

**Review by JavaSaBr — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    public boolean removeSubscriber(@NotNull MqttClient client) {
+        return subscribers.removeIfConvertedInWriteLock(client, SingleSubscriber::getMqttClient, Objects::equals);
+    }
+
+    public int size() {
+        //noinspection ConstantConditions
+        return subscribers.getInReadLock(Collection::size);
+    }
+
+    private static @NotNull SingleSubscriber next(@NotNull Array<SingleSubscriber> subscribers, int current) {
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/SharedSubscriber.java]: a static method should be at the top of this class

File: src/main/java/com/ss/mqtt/broker/util/TopicUtils.java [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+        if (isInvalidTopicFilter(topicFilter)) {
+            return INVALID_TOPIC_FILTER;
+        } else if (isShared(topicFilter)) {
+            return newSharedTopicFilter(topicFilter);
+        } else {
+            return new TopicFilter(topicFilter);
+        }
+    }
+
+    public static @NotNull List<String> splitTopic(@NotNull String topic) {
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/util/TopicUtils.java]: I prefer to use Array<> everywhere

**Review by JavaSaBr — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/mqtt/broker/model/topic/AbstractTopic.java [hunk 19]
```diff
@@ -15,41 +16,32 @@
...
-    static void checkTopic(@NotNull String topic) {
-        if (topic.length() == 0) {
-            throw new IllegalArgumentException("Topic has zero length.");
-        } else if (topic.contains("//") || topic.startsWith("/") || topic.endsWith("/")) {
-            throw new IllegalArgumentException("Topic has zero length level: " + topic);
-        }
-    }
-
-    private final @NotNull String[] segments;
+    private final @NotNull Array<String> segments;
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/topic/AbstractTopic.java]: I think any topic is immutable so better to use native array instead of any collection here

File: src/main/java/com/ss/mqtt/broker/util/TopicUtils.java [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+        if (isInvalidTopicFilter(topicFilter)) {
+            return INVALID_TOPIC_FILTER;
+        } else if (isShared(topicFilter)) {
+            return buildSharedTopicFilter(topicFilter);
+        } else {
+            return new TopicFilter(topicFilter);
+        }
+    }
+
+    public static @NotNull Array<String> splitTopic(@NotNull String topic) {
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/util/TopicUtils.java]: do we have any case when we need to have collection of segments? not a native array?

**Review by JavaSaBr — CHANGES_REQUESTED** [review]

File: src/main/java/com/ss/mqtt/broker/model/topic/AbstractTopic.java (hunk 19 above)
Comment (crazyrokr) [reply src/main/java/com/ss/mqtt/broker/model/topic/AbstractTopic.java]: Fixed

File: gradle/wrapper/gradle-wrapper.properties [hunk 21]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.0-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.4.1-all.zip
```
Comment (JavaSaBr) [inline gradle/wrapper/gradle-wrapper.properties]: wrong version of gradle

File: src/main/java/com/ss/mqtt/broker/model/SingleSubscriber.java [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+import org.jetbrains.annotations.NotNull;
+
+@ToString
+@EqualsAndHashCode(of = "mqttClient")
+public class SingleSubscriber implements Subscriber {
+
+    private final @Getter @NotNull MqttClient mqttClient;
+    private final @NotNull SubscribeTopicFilter subscribe;
+
+    public SingleSubscriber(@NotNull MqttClient mqttClient, @NotNull SubscribeTopicFilter subscribe) {
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/SingleSubscriber.java]: you can use lombok here

File: src/main/java/com/ss/mqtt/broker/model/topic/TopicSubscribers.java [hunk 23]
```diff
@@ -65,57 +161,73 @@ private static boolean removeDuplicateWithLowerQoS(
...
     }
 
-    public @NotNull Array<Subscriber> matches(@NotNull TopicName topicName) {
-        var resultArray = Array.ofType(Subscriber.class);
+    private boolean tryToRemoveSubscriber(@NotNull TopicFilter topicFilter, @NotNull MqttClient mqttClient) {
+        var removed = false;
+        var subscribers = getSubscribers();
+        if (subscribers != null) {
+            //noinspection ConstantConditions
+            removed = subscribers.getInWriteLock(
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/topic/TopicSubscribers.java]: you can return result here

File: src/main/java/com/ss/mqtt/broker/model/topic/TopicSubscribers.java [hunk 24]
```diff
@@ -65,57 +161,73 @@ private static boolean removeDuplicateWithLowerQoS(
...
+                topicFilter,
+                mqttClient,
+                TopicSubscribers::removeSubscriber
+            );
+        }
+        return removed;
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/model/topic/TopicSubscribers.java]: you can return here just 'false'

File: src/main/java/com/ss/mqtt/broker/util/SubscriberUtils.java [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+
+    private static boolean isSharedSubscriber(@NotNull Subscriber subscriber) {
+        return subscriber instanceof SharedSubscriber;
+    }
+
+    private static boolean isSingleSubscriber(@NotNull Subscriber subscriber) {
+        return subscriber instanceof SingleSubscriber;
+    }
+
+    public static boolean sharedSubscriberWithGroup(@NotNull String group, @NotNull Subscriber subscriber) {
```
Comment (JavaSaBr) [inline src/main/java/com/ss/mqtt/broker/util/SubscriberUtils.java]: maybe 'isSharedSubscriberWithGroup'

**Review by JavaSaBr — CHANGES_REQUESTED** [review]

**Review by JavaSaBr — APPROVED** [review]
